bson/bsoncodec/default_value_encoders.go

```diff
@@ -104,7 +104,7 @@ func (dve DefaultValueEncoders) RegisterDefaultEncoders(rb *RegistryBuilder) {
 		RegisterDefaultEncoder(reflect.Map, defaultMapCodec).
 		RegisterDefaultEncoder(reflect.Slice, defaultSliceCodec).
 		RegisterDefaultEncoder(reflect.String, defaultStringCodec).
-		RegisterDefaultEncoder(reflect.Struct, defaultStructCodec).
+		RegisterDefaultEncoder(reflect.Struct, newDefaultStructCodec()).
```

Comment:
I don't have a sense of whether the cache actually has any noticeable affect, but if it does. But I'm curious if there is a potential concern for a performance regression. Before, if you are using the default encoder/decoders, the cache was shared between encoding and decoding. Though once they're cached you won't need to derive the struct description again, so probably not a big concern.

If in the unlikely event this is a concern, perhaps we could keep the shared cache for encoding/decoding by default, and separate them when using a custom registry?

Reply:
Yeah, I came across this as well. I'm not too concerned because, as you said, we'll still cache after one encode and one decode. If this does become a problem, we can try to optimize by adding a new function similar to `RegisterDefaultEncoders`/`RegisterDefaultDecoders` that registers codecs like `StructCodec` that are shared during encoding and decoding. I filed GODRIVER-1759 to look into this, but I think we can merge this work as-is and look at that later.